### PR TITLE
Improve filtering in Snapshot#addedFiles.

### LINKED ISSUE #337: Can BaseSnapshot#cacheChanges() be optimized ?
@shardulm94 and me were discussing this yesterday.

In `org.apache.iceberg.BaseSnapshot#cacheChanges()` we are only accumulating manifest entries which were added or deleted by current snapshot. Since the manifest file  also contains the snapshot id of the snapshot which created it, it seems we can filter manifests by the current snapshot before we traverse their manifest entries. 

My understanding is that a `ManifestFile.snapshotId()` will always match the snapshot ids in manifest-entries if these entries are adds or deletes, isn't?

### PR DESCRIPTION
**Pull request by rdblue**
This updates `addedFiles` and `deletedFiles` to scan only the manifests that were created by a snapshot to improve performance.

Fixes #337.

### CODE REVIEW

**Comment by rdblue** [comment]
@rdsr, this fixes the performance problem you hit in #315.

File: core/src/main/java/org/apache/iceberg/BaseSnapshot.java [hunk 1]
```diff
@@ -152,32 +154,35 @@ public String manifestListLocation() {
   }
 
   private void cacheChanges() {
```
Comment (rdsr) [inline core/src/main/java/org/apache/iceberg/BaseSnapshot.java]: This fixes #337 too

File: core/src/main/java/org/apache/iceberg/BaseSnapshot.java [hunk 2]
```diff
@@ -152,32 +154,35 @@ public String manifestListLocation() {
...
-          }
+    ImmutableList.Builder<DataFile> adds = ImmutableList.builder();
+    ImmutableList.Builder<DataFile> deletes = ImmutableList.builder();
+
+    // read only manifests that were created by this snapshot
+    Iterable<ManifestFile> changedManifests = Iterables.filter(manifests(),
+        manifest -> Objects.equal(manifest.snapshotId(), snapshotId));
+    try (CloseableIterable<ManifestEntry> entries = new ManifestGroup(ops, changedManifests)
+        .ignoreExisting()
+        .select("file_path", "file_format", "partition", "record_count", "file_size_in_bytes")
```
Comment (rdsr) [inline core/src/main/java/org/apache/iceberg/BaseSnapshot.java]: nit: might we worthwhile to use `ManifestReader#CHANGE_COLUNNS` here.

File: core/src/main/java/org/apache/iceberg/ManifestReader.java [hunk 3]
```diff
@@ -161,16 +161,20 @@ private void cacheChanges() {
...
-    for (ManifestEntry entry : entries(fileSchema.select(CHANGE_COLUNNS))) {
-      switch (entry.status()) {
-        case ADDED:
-          adds.add(entry.copyWithoutStats());
-          break;
-        case DELETED:
-          deletes.add(entry.copyWithoutStats());
-          break;
-        default:
+    try (CloseableIterable<ManifestEntry> entries = entries(fileSchema.select(CHANGE_COLUNNS))) {
```
Comment (rdsr) [inline core/src/main/java/org/apache/iceberg/ManifestReader.java]: What is the purpose for `ManifestReader#cacheChanges()`. It seems previously it was used in `Basesnapshot#cacheChanges()` but now is it even used anywhere? .

**Comment by rdsr** [comment]
This fixes #337 too

File: core/src/main/java/org/apache/iceberg/ManifestReader.java (hunk 3 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/ManifestReader.java]: Also `ManifestReader#[addedFiles(), deletedFiles()]`

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/ManifestReader.java]: The intent is the same as for Snapshot: to make the added and deleted files easily accessible. Manifests are `Iterable<DataFile>`, and don't expose the internal `ManifestEntry` data. That means we need to make the added and deleted files accessible through the API somehow.

**Comment by rdblue** [comment]
Thanks for reviewing, @rdsr! I'm going to merge this.

**Pull request merged**